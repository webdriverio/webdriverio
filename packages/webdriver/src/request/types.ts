import type { Options } from '@wdio/types'

export interface WebDriverResponse {
    value: any
    /**
     * error case
     * https://w3c.github.io/webdriver/webdriver-spec.html#dfn-send-an-error
     */
    error?: string
    message?: string
    stacktrace?: string

    /**
     * JSONWP property
     */
    status?: number
    sessionId?: string
}

export type RequestLibResponse = Options.RequestLibResponse
export type RequestOptions = Omit<Options.WebDriver, 'capabilities'>
