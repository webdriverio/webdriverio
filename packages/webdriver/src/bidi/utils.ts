import { BASE_64_REGEX, BASE_64_SAFE_STRING_TO_PROCESS_LENGTH } from '../constants.js'

export function isBase64Safe(str: string): boolean {
    if (typeof str !== 'string') {
        return false
    }

    if (str.length === 0) {
        return true
    }

    if (str.length % 4 !== 0) {
        return false
    }

    const length = str.length

    // Calculate number of digits in the length (e.g., 1000000 => 7)
    const digitCount = length.toString().length

    // Only chunk if length > BASE_64_SAFE_STRING_TO_PROCESS_LENGTH
    if (length > BASE_64_SAFE_STRING_TO_PROCESS_LENGTH) {
        const chunkSize = Math.floor(length / digitCount / 4) * 4
        for (let i = 0; i < length; i += chunkSize) {
            const chunk = str.slice(i, i + chunkSize)
            if (!BASE_64_REGEX.test(chunk)) {
                return false
            }
        }
        return true
    }

    // For shorter strings, validate whole thing
    return BASE_64_REGEX.test(str)
}

