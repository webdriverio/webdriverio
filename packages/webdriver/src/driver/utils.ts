import os from 'node:os'
import fs from 'node:fs'
import path from 'node:path'
import cp from 'node:child_process'

import decamelize from 'decamelize'
import logger from '@wdio/logger'
import { getChromePath } from 'chrome-launcher'
import {
    install, canDownload, resolveBuildId, detectBrowserPlatform, Browser, ChromeReleaseChannel,
    computeExecutablePath, type InstallOptions
} from '@puppeteer/browsers'
import type { EdgedriverParameters } from 'edgedriver'
import type { Capabilities, Options } from '@wdio/types'

import { DEFAULTS } from '../constants.js'

const log = logger('webdriver')
const EXCLUDED_PARAMS = ['version', 'help']

export function parseParams(params: EdgedriverParameters) {
    return Object.entries(params)
        .filter(([key,]) => !EXCLUDED_PARAMS.includes(key))
        .map(([key, val]) => {
            if (typeof val === 'boolean' && !val) {
                return ''
            }
            const vals = Array.isArray(val) ? val : [val]
            return vals.map((v) => `--${decamelize(key, { separator: '-' })}${typeof v === 'boolean' ? '' : `=${v}`}`)
        })
        .flat()
        .filter(Boolean)
}

export function getLocalChromePath() {
    try {
        const chromePath = getChromePath()
        return chromePath
    } catch (err: unknown) {
        return
    }
}

export function getBuildIdByPath(chromePath?: string) {
    if (!chromePath) {
        return
    } else if (os.platform() === 'win32') {
        const versionPath = path.dirname(chromePath)
        const contents = fs.readdirSync(versionPath)
        const versions = contents.filter(a => /^[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+$/g.test(a))

        // returning oldest in case there is an updated version and chrome still hasn't relaunched
        const oldest = versions.sort((a: string, b: string) => a > b ? -1 : 1)[0]
        return oldest
    }

    const versionString = cp.execSync(`"${chromePath}" --version`).toString()
    return versionString.split(' ').pop()?.trim()
}

let lastTimeCalled = Date.now()
export const downloadProgressCallback = (artifact: string, downloadedBytes: number, totalBytes: number) => {
    if (Date.now() - lastTimeCalled < 1000) {
        return
    }
    const percentage = ((downloadedBytes / totalBytes) * 100).toFixed(2)
    log.info(`Downloading ${artifact} ${percentage}%`)
    lastTimeCalled = Date.now()
}

export async function setupChrome(caps: Capabilities.Capabilities, cacheDir: string) {
    const platform = detectBrowserPlatform()
    if (!platform) {
        throw new Error('The current platform is not supported.')
    }

    if (!caps.browserVersion) {
        const executablePath = getLocalChromePath()!
        const tag = getBuildIdByPath(executablePath)

        /**
         * verify that we have a valid Chrome browser installed
         */
        if (tag) {
            return {
                platform,
                executablePath,
                buildId: await resolveBuildId(Browser.CHROME, platform, tag)
            }
        }
    }

    /**
     * otherwise download provided Chrome browser version or "stable"
     */
    const tag = caps.browserVersion || ChromeReleaseChannel.STABLE
    const buildId = await resolveBuildId(Browser.CHROME, platform, tag)
    const installOptions: InstallOptions & { unpack?: true } = {
        unpack: true,
        cacheDir,
        platform,
        buildId,
        browser: Browser.CHROME,
        downloadProgressCallback: (downloadedBytes, totalBytes) => downloadProgressCallback('Chrome', downloadedBytes, totalBytes)
    }
    const isCombinationAvailable = await canDownload(installOptions)
    if (!isCombinationAvailable) {
        throw new Error(`Couldn't find a matching Chrome browser for tag "${buildId}" on platform "${platform}"`)
    }

    log.info(`Setting up Chrome v${buildId}`)
    await install(installOptions)
    const executablePath = computeExecutablePath(installOptions)
    return { executablePath, buildId, platform }
}

/**
 * helper method to determine if we need to start a browser driver
 * which is: whenever the user has set connection options that differ from
 * the default, or a port is set
 */
export function definesRemoteDriver(options: Options.WebDriver) {
    return Boolean(
        (options.protocol && options.protocol !== DEFAULTS.protocol.default) ||
        (options.hostname && options.hostname !== DEFAULTS.hostname.default) ||
        Boolean(options.port) ||
        (options.path && options.path !== DEFAULTS.path.default)
    )
}
