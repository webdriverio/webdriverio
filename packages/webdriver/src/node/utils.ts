import logger from '@wdio/logger'

const log = logger('webdriver')

/**
 * Kill the driver process after `deleteSession` command is done
 * @param capabilities - the capabilities of the driver
 * @param shutdownDriver - whether to shutdown the driver
 */
export function killDriverProcess(capabilities: WebdriverIO.Capabilities, shutdownDriver: boolean) {
    if (shutdownDriver && 'wdio:driverPID' in capabilities && capabilities['wdio:driverPID']) {
        log.info(`Kill driver process with PID ${capabilities['wdio:driverPID']}`)
        try {
            const killedSuccessfully = process.kill(capabilities['wdio:driverPID'], 'SIGKILL')
            if (!killedSuccessfully) {
                log.warn('Failed to kill driver process, manually clean-up might be required')
            }
        } catch (err) {
            log.warn('Failed to kill driver process', err)
        }

        setTimeout(() => {
            /**
             * clear up potential leaked TLS Socket handles
             * see https://github.com/puppeteer/puppeteer/pull/10667
             */
            for (const handle of process._getActiveHandles()) {
                if (handle.servername && handle.servername.includes('edgedl.me')) {
                    handle.destroy()
                }
            }
        }, 10)
    }
}
