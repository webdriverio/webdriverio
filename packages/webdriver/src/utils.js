import logger from 'wdio-logger'

import command from './command'
import WebDriverProtocol from '../protocol/webdriver.json'
import MJsonWProtocol from '../protocol/mjsonwp.json'
import JsonWProtocol from '../protocol/jsonwp.json'
import AppiumProtocol from '../protocol/appium.json'

const log = logger('webdriver')

/**
 * check if WebDriver requests was successful
 * @param  {Object}  body  body payload of response
 * @return {Boolean}       true if request was successful
 */
export function isSuccessfulResponse ({ body, statusCode } = {}) {
    /**
     * response contains a body
     */
    if (!body || typeof body.value === 'undefined') {
        log.debug('request failed due to missing body')
        return false
    }

    /**
     * ignore failing element request to enable lazy loading capability
     */
    if (body.status && body.status === 7 && body.value.message && body.value.message.startsWith('no such element')) {
        return true
    }

    /**
     * if it has a status property, it should be 0
     * (just here to stay backwards compatible to the jsonwire protocol)
     */
    if (body.status && body.status !== 0) {
        log.debug(`request failed due to status ${body.status}`)
        return false
    }

    const hasErrorResponse = body.value && (body.value.error || body.value.stackTrace || body.value.stacktrace)

    /**
     * check status code
     */
    if (statusCode === 200 && !hasErrorResponse) {
        return true
    }

    /**
     * if an element was not found we don't flag it as failed request because
     * we lazy load it
     */
    if (statusCode === 404 && body.value && body.value.error === 'no such element') {
        return true
    }

    /**
     * that has no error property (Appium only)
     */
    if (hasErrorResponse) {
        log.debug('request failed due to response error:', body.value.error)
        return false
    }

    return true
}

/**
 * checks if command argument is valid according to specificiation
 *
 * @param  {*}       arg           command argument
 * @param  {Object}  expectedType  parameter type (e.g. `number`, `string[]` or `(number|string)`)
 * @return {Boolean}               true if argument is valid
 */
export function isValidParameter (arg, expectedType) {
    let shouldBeArray = false

    if (expectedType.slice(-2) === '[]') {
        expectedType = expectedType.slice(0, -2)
        shouldBeArray = true
    }

    /**
     * check type of each individual array element
     */
    if (shouldBeArray) {
        if (!Array.isArray(arg)) {
            return false
        }
    } else {
        /**
         * transform to array to have a unified check
         */
        arg = [arg]
    }

    for (const argEntity of arg) {
        if (!(typeof argEntity).match(expectedType)) {
            return false
        }
    }

    return true
}

/**
 * creates the base prototype for the webdriver monad
 */
export function getPrototype (isW3C) {
    const prototype = {}
    const ProtocolCommands = Object.assign(
        isW3C ? WebDriverProtocol : JsonWProtocol,
        MJsonWProtocol,
        AppiumProtocol
    )

    for (const [endpoint, methods] of Object.entries(ProtocolCommands)) {
        for (const [method, commandData] of Object.entries(methods)) {
            prototype[commandData.command] = { value: command(method, endpoint, commandData) }
        }
    }

    return prototype
}

/**
 * get command call structure
 * (for logging purposes)
 */
export function commandCallStructure (commandName, args) {
    const callArgs = args.map((arg) => {
        if (typeof arg === 'string') {
            arg = `"${arg}"`
        } else if (typeof arg === 'function') {
            arg = '<fn>'
        } else if (arg === null) {
            arg = 'null'
        } else if (typeof arg === 'object') {
            arg = '<object>'
        } else if (typeof arg === 'undefined') {
            arg = typeof arg
        }

        return arg
    }).join(', ')

    return `${commandName}(${callArgs})`
}
