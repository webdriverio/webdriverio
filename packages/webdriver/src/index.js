import logger from '@wdio/logger'
import merge from 'lodash.merge'
import { validateConfig } from '@wdio/config'

import webdriverMonad from './monad'
import WebDriverRequest from './request'
import { DEFAULTS } from './constants'
import { getPrototype, environmentDetector } from './utils'

import WebDriverProtocol from '../protocol/webdriver.json'
import JsonWProtocol from '../protocol/jsonwp.json'
import MJsonWProtocol from '../protocol/mjsonwp.json'
import AppiumProtocol from '../protocol/appium.json'
import ChromiumProtocol from '../protocol/chromium.json'

export default class WebDriver {
    static async newSession (options = {}, modifier, userPrototype = {}, commandWrapper) {
        const params = validateConfig(DEFAULTS, options)
        logger.setLevel('webdriver', params.logLevel)

        /**
         * the user could have passed in either w3c style or jsonwp style caps
         * and we want to pass both styles to the server, which means we need
         * to check what style the user sent in so we know how to construct the
         * object for the other style
         */
        const [w3cCaps, jsonwpCaps] = params.capabilities && params.capabilities.alwaysMatch
            /**
             * in case W3C compliant capabilities are provided
             */
            ? [params.capabilities, params.capabilities.alwaysMatch]
            /**
             * otherwise assume they passed in jsonwp-style caps (flat object)
             */
            : [{ alwaysMatch: params.capabilities, firstMatch: [{}] }, params.capabilities]

        const sessionRequest = new WebDriverRequest(
            'POST',
            '/session',
            {
                capabilities: w3cCaps, // W3C compliant
                desiredCapabilities: jsonwpCaps // JSONWP compliant
            }
        )

        const response = await sessionRequest.makeRequest(params)

        /**
         * save original set of capabilities to allow to request the same session again
         * (e.g. for reloadSession command in WebdriverIO)
         */
        params.requestedCapabilities = { w3cCaps, jsonwpCaps }

        /**
         * save actual receveived session details
         */
        params.capabilities = response.value.capabilities || response.value

        /**
         * apply mobile flags to driver scope
         */
        const { isW3C, isMobile, isIOS, isAndroid, isChrome } = environmentDetector(params.capabilities)
        const environmentFlags = {
            isW3C: { value: isW3C },
            isMobile: { value: isMobile },
            isIOS: { value: isIOS },
            isAndroid: { value: isAndroid },
            isChrome: { value: isChrome }
        }

        const protocolCommands = getPrototype({ isW3C, isMobile, isIOS, isAndroid, isChrome })
        const prototype = merge(protocolCommands, environmentFlags, userPrototype)
        const monad = webdriverMonad(params, modifier, prototype)
        return monad(response.value.sessionId || response.sessionId, commandWrapper)
    }

    /**
     * allows user to attach to existing sessions
     */
    static attachToSession (options = {}, modifier, userPrototype = {}, commandWrapper) {
        if (typeof options.sessionId !== 'string') {
            throw new Error('sessionId is required to attach to existing session')
        }

        logger.setLevel('webdriver', options.logLevel)

        options.capabilities = options.capabilities || {}
        options.isW3C = options.isW3C || true
        const prototype = Object.assign(getPrototype(options.isW3C), userPrototype)
        const monad = webdriverMonad(options, modifier, prototype)
        return monad(options.sessionId, commandWrapper)
    }

    static get WebDriver () {
        return WebDriver
    }
    static get DEFAULTS () {
        return DEFAULTS
    }

    /**
     * Protocols
     */
    static get WebDriverProtocol () {
        return WebDriverProtocol
    }
    static get JsonWProtocol () {
        return JsonWProtocol
    }
    static get MJsonWProtocol () {
        return MJsonWProtocol
    }
    static get AppiumProtocol () {
        return AppiumProtocol
    }
    static get ChromiumProtocol () {
        return ChromiumProtocol
    }
}

/**
 * Helper methods consumed by webdriverio package
 */
export {
    webdriverMonad,
    getPrototype
}
