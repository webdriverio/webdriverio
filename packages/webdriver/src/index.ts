import path from 'path'
import logger from '@wdio/logger'

import { webdriverMonad, sessionEnvironmentDetector } from '@wdio/utils'
import { validateConfig } from '@wdio/config'

import { DEFAULTS } from './constants'
import { startWebDriverSession, getPrototype, getEnvironmentVars } from './utils'
import type {
    Options, Client, AttachOptions, SessionFlags,
    DesiredCapabilities
} from './types'

const log = logger('webdriver')

export default class WebDriver {
    static async newSession (
        options: Options = {},
        modifier?: (...args: any[]) => any,
        userPrototype = {},
        customCommandWrapper?: (...args: any[]) => any
    ): Promise<Client> {
        const params = validateConfig(DEFAULTS, options)

        if (!options.logLevels || !options.logLevels.webdriver) {
            logger.setLevel('webdriver', params.logLevel!)
        }

        /**
         * if the server responded with direct connect information, update the
         * params to speak directly to the appium host instead of a load
         * balancer (see https://github.com/appium/python-client#direct-connect-urls
         * for example). But only do this if the user has enabled this
         * behavior in the first place.
         */
        const { directConnectProtocol, directConnectHost, directConnectPort, directConnectPath } = params
        if (directConnectProtocol && directConnectHost && directConnectPort && (directConnectPath || directConnectPath === '')) {
            log.info('Found direct connect information in new session response. ' +
                `Will connect to server at ${directConnectProtocol}://${directConnectHost}:${directConnectPort}/${directConnectPath}`)
            params.protocol = directConnectProtocol
            params.hostname = directConnectHost
            params.port = directConnectPort
            params.path = directConnectPath
        }

        /**
         * Store all log events in a file
         */
        if (params.outputDir) {
            process.env.WDIO_LOG_PATH = path.join(params.outputDir, 'wdio.log')
        }

        const { sessionId, capabilities } = await startWebDriverSession(params)
        const environment = sessionEnvironmentDetector({ capabilities, requestedCapabilities: params.requestedCapabilities })
        const environmentPrototype = getEnvironmentVars(environment)
        const protocolCommands = getPrototype(environment)
        const prototype = { ...protocolCommands, ...environmentPrototype, ...userPrototype }

        const monad = webdriverMonad(params, modifier, prototype)
        return monad(sessionId, customCommandWrapper)
    }

    /**
     * allows user to attach to existing sessions
     */
    static attachToSession (
        options?: AttachOptions,
        modifier?: (...args: any[]) => any,
        userPrototype = {},
        commandWrapper?: (...args: any[]) => any
    ): Client {
        if (!options || typeof options.sessionId !== 'string') {
            throw new Error('sessionId is required to attach to existing session')
        }

        // logLevel can be undefined in watch mode when SIGINT is called
        if (options.logLevel !== undefined) {
            logger.setLevel('webdriver', options.logLevel)
        }

        options.capabilities = options.capabilities || {}
        options.isW3C = options.isW3C === false ? false : true

        const environmentPrototype = getEnvironmentVars(options as Partial<SessionFlags>)
        const protocolCommands = getPrototype(options as Partial<SessionFlags>)
        const prototype = { ...protocolCommands, ...environmentPrototype, ...userPrototype }
        const monad = webdriverMonad(options, modifier, prototype)
        return monad(options.sessionId, commandWrapper)
    }

    /**
     * Changes The instance session id and browser capabilities for the new session
     * directly into the passed in browser object
     *
     * @param   {Object} instance  the object we get from a new browser session.
     * @returns {string}           the new session id of the browser
    */
    static async reloadSession (instance: Client) {
        const params: Options = {
            ...instance.options,
            capabilities: instance.requestedCapabilities as DesiredCapabilities
        }
        const { sessionId, capabilities } = await startWebDriverSession(params)
        instance.sessionId = sessionId
        instance.capabilities = capabilities
        return sessionId
    }

    static get WebDriver () {
        return WebDriver
    }
}

/**
 * Helper methods consumed by webdriverio package
 */
export { getPrototype, DEFAULTS }
