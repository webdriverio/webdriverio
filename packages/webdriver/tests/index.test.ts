import path from 'path'
import gotMock from 'got'
import logger from '@wdio/logger'
import * as wdioUtils from '@wdio/utils'

import WebDriver, { getPrototype, DEFAULTS } from '../src'
import { DesiredCapabilities, Client } from '../src/types'

const got = gotMock as unknown as jest.Mock
const sessionEnvironmentDetector = wdioUtils.sessionEnvironmentDetector as jest.Mock

const sessionOptions = {
    protocol: 'http',
    hostname: 'localhost',
    port: 4444,
    path: '/',
    sessionId: 'foobar'
}

interface TestClient extends Client {
    getUrl (): string
    getApplicationCacheStatus (): void
    takeElementScreenshot (): void
    getDeviceTime (): void
}

describe('WebDriver', () => {
    test('exports getPrototype, DEFAULTS', () => {
        expect(typeof getPrototype).toBe('function')
        expect(typeof DEFAULTS).toBe('object')
    })
    describe('newSession', () => {
        it('should allow to create a new session using jsonwire caps', async () => {
            const testDirPath = './logs'
            await WebDriver.newSession({
                path: '/',
                outputDir: testDirPath,
                capabilities: { browserName: 'firefox' }
            })

            const url = got.mock.calls[0][0]
            const req = got.mock.calls[0][1]
            expect(url.pathname).toBe('/session')
            expect(req.json).toEqual({
                capabilities: {
                    alwaysMatch: { browserName: 'firefox' },
                    firstMatch: [{}]
                },
                desiredCapabilities: { browserName: 'firefox' }
            })
            expect(process.env.WDIO_LOG_PATH).toEqual(path.join(testDirPath, 'wdio.log'))
        })

        it('should allow to create a new session using w3c compliant caps', async () => {
            await WebDriver.newSession({
                path: '/',
                capabilities: {
                    alwaysMatch: { browserName: 'firefox' },
                    firstMatch: [{}]
                }
            })

            const url = got.mock.calls[0][0]
            const req = got.mock.calls[0][1]
            expect(url.pathname).toBe('/session')
            expect(req.json).toEqual({
                capabilities: {
                    alwaysMatch: { browserName: 'firefox' },
                    firstMatch: [{}]
                },
                desiredCapabilities: { browserName: 'firefox' }
            })

            expect(sessionEnvironmentDetector.mock.calls)
                .toMatchSnapshot()
        })

        it('should allow to use Appium direct connect functionality', async () => {
            await WebDriver.newSession({
                directConnectProtocol: 'https',
                directConnectHost: 'foobar',
                directConnectPort: 1234,
                directConnectPath: '/foo/bar',
                path: '/',
                capabilities: {
                    alwaysMatch: { browserName: 'firefox' },
                    firstMatch: [{}]
                }
            })

            const url = got.mock.calls[0][0]
            expect(url.href).toEqual('https://foobar:1234/foo/bar/session')
        })

        it('should be possible to skip setting logLevel', async () => {
            await WebDriver.newSession({
                capabilities: { browserName: 'chrome' },
                logLevel: 'info',
                logLevels: { webdriver: 'silent' }
            })

            expect(logger.setLevel).not.toBeCalled()
        })

        it('should be possible to set logLevel', async () => {
            await WebDriver.newSession({
                capabilities: { browserName: 'chrome' },
                logLevel: 'info'
            })

            expect(logger.setLevel).toBeCalled()
        })

        it('propagates capabilities and requestedCapabilities', async () => {
            const browser = await WebDriver.newSession({
                path: '/',
                capabilities: { browserName: 'firefox' }
            })
            expect((browser.capabilities as DesiredCapabilities).browserName).toBe('mockBrowser')
            expect((browser.requestedCapabilities as DesiredCapabilities).browserName).toBe('firefox')
        })
    })

    describe('attachToSession', () => {
        it('should allow to attach to existing session', async () => {
            const client = WebDriver.attachToSession({ ...sessionOptions, logLevel: 'info' }) as TestClient
            await client.getUrl()
            const url = got.mock.calls[0][0]
            expect(url.href).toBe('http://localhost:4444/session/foobar/url')
            expect(logger.setLevel).toBeCalled()
        })

        it('should allow to attach to existing session2', async () => {
            const client = WebDriver.attachToSession({ ...sessionOptions }) as TestClient
            await client.getUrl()
            const url = got.mock.calls[0][0]
            expect(url.href).toBe('http://localhost:4444/session/foobar/url')
            expect(logger.setLevel).not.toBeCalled()
        })

        it('should allow to attach to existing session - W3C', async () => {
            const client = WebDriver.attachToSession({ ...sessionOptions }) as TestClient
            await client.getUrl()

            expect(client.isChrome).toBeFalsy()
            expect(client.isMobile).toBeFalsy()
            expect(client.isSauce).toBeFalsy()
            expect(client.getApplicationCacheStatus).toBeFalsy()
            expect(client.takeElementScreenshot).toBeTruthy()
            expect(client.getDeviceTime).toBeFalsy()
        })

        it('should allow to attach to existing session - non W3C', async () => {
            const client = WebDriver.attachToSession({ ...sessionOptions,
                isW3C: false,
                isSauce: true,
            }) as TestClient

            await client.getUrl()

            expect(client.isSauce).toBe(true)
            expect(client.getApplicationCacheStatus).toBeTruthy()
            expect(client.takeElementScreenshot).toBeFalsy()
            expect(client.getDeviceTime).toBeFalsy()
        })

        it('should allow to attach to existing session - mobile', async () => {
            const client = WebDriver.attachToSession({ ...sessionOptions,
                isChrome: true,
                isMobile: true
            }) as TestClient

            await client.getUrl()

            expect(client.isChrome).toBe(true)
            expect(client.isMobile).toBe(true)
            expect(client.getApplicationCacheStatus).toBeTruthy()
            expect(client.takeElementScreenshot).toBeTruthy()
            expect(client.getDeviceTime).toBeTruthy()
        })

        it('it should propagate all environment flags', () => {
            const client = WebDriver.attachToSession({ ...sessionOptions,
                isW3C: false,
                isMobile: false,
                isIOS: false,
                isAndroid: false,
                isChrome: false,
                isSauce: false
            })
            expect(client.isW3C).toBe(false)
            expect(client.isMobile).toBe(false)
            expect(client.isIOS).toBe(false)
            expect(client.isAndroid).toBe(false)
            expect(client.isChrome).toBe(false)
            expect(client.isSauce).toBe(false)

            const anotherClient = WebDriver.attachToSession({ ...sessionOptions,
                isW3C: true,
                isMobile: true,
                isIOS: true,
                isAndroid: true,
                isChrome: true,
                isSauce: true
            })
            expect(anotherClient.isW3C).toBe(true)
            expect(anotherClient.isMobile).toBe(true)
            expect(anotherClient.isIOS).toBe(true)
            expect(anotherClient.isAndroid).toBe(true)
            expect(anotherClient.isChrome).toBe(true)
            expect(anotherClient.isSauce).toBe(true)
        })

        it('should fail attaching to session if sessionId is not given', () => {
            // @ts-ignore
            expect(() => WebDriver.attachToSession({}))
                .toThrow(/sessionId is required/)
        })
    })

    describe('reloadSession', () => {
        it('should reload session', async () => {
            const session = await WebDriver.newSession({
                path: '/',
                capabilities: { browserName: 'firefox' }
            })
            await WebDriver.reloadSession(session)
            expect(got.mock.calls).toHaveLength(2)
        })
    })

    it('ensure that WebDriver interface exports protocols and other objects', () => {
        expect(WebDriver.WebDriver).not.toBe(undefined)
    })

    afterEach(() => {
        (logger.setLevel as jest.Mock).mockClear()
        got.mockClear()
        sessionEnvironmentDetector.mockClear()
    })
})
