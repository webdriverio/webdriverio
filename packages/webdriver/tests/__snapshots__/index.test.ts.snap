// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebDriver newSession should allow to create a new session using w3c compliant caps 1`] = `
Array [
  Array [
    Object {
      "capabilities": Object {
        "browserName": "mockBrowser",
        "browserVersion": "1234",
        "platformName": "node",
        "setWindowRect": true,
      },
      "requestedCapabilities": Object {
        "alwaysMatch": Object {
          "browserName": "firefox",
        },
        "firstMatch": Array [
          Object {},
        ],
      },
    },
  ],
]
`;
